### PR TITLE
Make sprites respect RenderLayers

### PR DESCRIPTION
**Pull request by kserz**
# Objective

Make sprites respect RenderLayers. Alternative to #4007.

## Solution

- Add ComputedVisibility to SpriteBundle, SpriteSheetBundle, Text2dBundle and sprites extract stage. Potentially useful for dropping sprites by AABB checks.
- At sprites queue stage use RenderLayers directly.
- As a side fix: sorting of extracted sprites is done once for all the views as it is not changed in between.

Compared to #4007:

- Pros: no performance penalty on many_sprites.
- Cons: some duplicate logic between visibility computation and sprites.

### CODE REVIEW

File: crates/bevy_text/src/text2d.rs
Comment (alice-i-cecile) [inline crates/bevy_text/src/text2d.rs:65]:
```suggestion
    pub computed_visibility: Visibility,
```

**Review by alice-i-cecile — COMMENTED** [review]
Nice work, thanks for the contribution. Frustum culling is not enabled in 2D still, correct?

I think this is a relatively important fix, and I like the code changes.

**Comment by alice-i-cecile** [comment]
Without any performance losses, I don't think this is a controversial change. @johanhelsing, could I get your review over here?

**Comment by alice-i-cecile** [comment]
>  On the other hand, ComputedVisibility here is global visibility, not per-view

I don't think this is sufficient. This will break things like split screen right?

>  I don’t remember off the top of my head if simply adding ComputedVisibility and an Aabb is sufficient for culling.

This would need to be an oriented bounding box, but basically yes.

> I think we should just do 2d- and 3d-specific culling code as 2d is so much simpler than 3d.

I agree; I think that the duplication is worth the performance gains and will allow us to iterate on 2D transform types more easily.

**Comment by kserz** [comment]
As far as I can tell, AABB's are not computed for sprites currently, so culling is likely not working.

I'm not sure about addition of ComputedVisibility as well. On the one hand, it is not a requirement for RenderLayers to work. On the other hand, all the sprites not present in any view will be dropped earlier. Currently by layers only in case of dynamic layers switching, I'm not sure whether it's useful. With the addition of AABB's that would be more beneficial to reduce the amount of extracted sprites and sorting time.

I can remove ComputedVisibility from sprites for now and leave it until there is a clear benefit. Any opinions?

**Comment by kserz** [comment]
> > On the other hand, ComputedVisibility here is global visibility, not per-view
> 
> I don't think this is sufficient. This will break things like split screen right?

Probably I'm not understanding this. Split screen is actually my goal for this change: support for RenderLayers should allow rendering different sets of sprites for different players(cameras). For meshes (3d & 2d) it is already working through ComputedVisibibilty. For sprites it will work by accounting layers directly, and ComputedVisibility will be used as a kind of optimization.

**Comment by alice-i-cecile** [comment]
Sorry, I was unclear. I was concerned that storing a single `ComputedVisibility` list would prevent us from performing frustum culling using it in cases where more than one viewport exists.

> I'm not sure about addition of ComputedVisibility as well. On the one hand, it is not a requirement for RenderLayers to work. On the other hand, all the sprites not present in any view will be dropped earlier

Sounds like we could split out adding `ComputedVisibility` into a separate PR with its own motivation. Could we see what this change set looks like without the `ComputedVisibility` changes in a new commit?

**Comment by superdump** [comment]
I need some more time to think about what the right choice is here but I think minimally just adding RenderLayer support by extracting it and using it when drawing stuff for each view should be ok.

**Review by alice-i-cecile — APPROVED** [review]
I'm very happy with this as a simple incremental improvement. Adding to the milestone; this flagship feature should work in 2D on release.

**Review by james7132 — APPROVED** [review]
No glaring issues seen here, though ideally we shouldn't extract any sprites that aren't visible.

**Comment by alice-i-cecile** [comment]
@superdump I'm waiting on you to merge this.

**Comment by cart** [comment]
Before merging, I'd like to compare this to the fully unified visibility inheritance / ComputedVisibility everywhere impl im working on over in #5146. (see my rationale there).

**Comment by cart** [comment]
We should definitely merge one of these impls for 0.8 though. Which one matters way less than including one of them.

**Comment by james7132** [comment]
Superseded by #5310.

**Pull request closed without merging**
